rclcpp/include/rclcpp/parameter_descriptor_wrapper.hpp

```diff
@@ -0,0 +1,102 @@
+#ifndef RCLCPP__PARAMETER_DESCRIPTOR_WRAPPER_HPP_
+#define RCLCPP__PARAMETER_DESCRIPTOR_WRAPPER_HPP_
+
+// Standard library includes
+#include <functional>
+#include <utility>
+#include <memory>
+
+// Additional ROS libraries needed
+#include "node.hpp"
+#include "rcl_interfaces/msg/describe_parameters.hpp"
+#include "rclcpp/node.hpp"
+
+namespace rclcpp
+{
+
+// Implments ParameterDesription class with builder design pattern
+class ParameterDescription
+{
+public:
+    // List of classes the builder manages
+    class FloatingPointDescription;
+
+    // Two constructors in case certain methods don't need to declare_parameters with a Type T in the form of a template
+    ParameterDescription(std::string name, std::uint8_t type, std::string description, std::string additional_constraints, bool read_only. bool dynamic_typing) : m_name{name}, parameter_descriptor.type{type}, m_description{description}, m_additional_constraints{additional_constraints}, read_only{read_only}, dyanmic_typing{m_dyanmic_typing}{};
+
+private:
+    // The main descriptor object we're meant to adjust
+    rcl_interfaces::msg::ParameterDescriptor parameter_descriptor = {};
+
+    // Copies all the information in ParameterDescriptor.msg - https://github.com/ros2/rcl_interfaces/blob/rolling/rcl_interfaces/msg/ParameterDescriptor.msg
+    std::string m_name;
+    std::string m_description;
+
+    // Parameter Constraints
+    std::string m_additional_constraints;
+    bool m_read_only;
+    bool m_dynamic_typing;
+
+};
+
+class ParameterDescription::FloatingPointDescription
+{
+public:
+    // Our Main build methods which will construct the base class
+    ParameterDescription build() const;
+
+    //Builder Method - Mains
+    FloatingPointDescription& SetName(std::string name);
+    FloatingPointDescription& SetType(std::uint8_t type);
+    FloatingPointDescription& SetDescription(std::string description);
+    FloatingPointDescription& SetAdditionalConstraints(std::string constraints);
+    FloatingPointDescription& SetReadOnly(bool read_only);
+    FloatingPointDescription& SetDynamicTyping(bool dynamic_typing);
+
+    // Need the current node in order to begin the configuraiton state forit via the declare_parameter function which setups up the Node
+    template<typename ParameterType>
+    FloatingPointDescription& DeclareParameter<ParameterType>(ParameterType default_value, std::shared_ptr<rclcpp::Node> required_node)
+{
+    required_node->declare_parameter<ParameterType>(m_name, default_value, parameter_descriptor);
+    return *this;
+}
+    // Extended build methods specific to this class
+    FloatingPointDescription& SetMin(float min);
+    FloatingPointDescription& SetMax(float max);
+    FloatingPointDescription& SetStep(float step);
+
```

Comment:
Before I roll out the next set of changes how would we like these three functions: `SetMin`, `SetMax`, and `SetStep` to act?  There are multiple aspects to consider when deciding between simplicity and modularity. 


Currently these three functions have no use, because the values they set are unused. In order to make them have purpose we would not only set their corresponding value, but also resize the type of range it's trying to target, if that range had not already been set. After we would then call `parameter_descriptor.floating_point_range.at(0).corresponding_function`. On the other hand we can opt for simplicity by removing the set functions and the values they alter all together, then in the functions that Set the Description Range we can use default values to remove possible arguments because the .msg field specifies bounds that we can have. On top of this discussion, should we also create Set functions for the int range, but with different names i.e `SetMinInteger` or should we aim to use overloading like `SetMin(float/int min)`